api/ping: add support for querying engine features

laurazard opened the pull request:
Supercedes/close: https://github.com/moby/moby/pull/48944

**- What I did**

With moving towards using the containerd image store (as opposed to the "legacy" graphdrivers), some new daemon features will only be implemented for the containerd image store.

A client wanting to make use of new features would find it useful to be able to know what features are available without having to attempt to use them, getting an error back, and having to do another roundtrip.

Address this by adding a new query parameter to the `/_ping` endpoint to allow clients to query supported features.

**- How I did it**

Adds a new query parameter to the (`GET`) Ping endpoint `/_ping?capabilities=1` that, if set, prompts the daemon to reply with a JSON representation of it's configured features.

Since clients will generally already by hitting the `/_ping` endpoint in order to negotiate API versions, this provides a way for clients to know which features are supported without extra roundtrips.

Also updated swagger and API docs.

**- How to verify it**

```console
$ curl -s --unix-socket /var/run/docker.sock http://localhost/_ping?capabilities=1 | jq
{
  "capabilities": {
    "_v": 1,
    "data": {
      "registry-client-auth": true
    }
  }
}
```

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section:
-->
```markdown changelog
Added `capabilities` query parameter to the `/_ping` endpoint to request daemon capabilities.
```

**- A picture of a cute animal (not mandatory but encouraged)**

<img width="594" alt="Screenshot 2024-11-21 at 12 14 16" src="https://github.com/user-attachments/assets/3fed59ec-6078-45e6-9082-6ca0407c6001">

laurazard committed:
api/ping: add support for querying engine capabilities

With moving towards using the containerd image store (as opposed to the
"legacy" graphdrivers), some new daemon features will only be
implemented for the containerd image store.

A client wanting to make use of new features would find it useful to be
able to know what features are available without having to attempt to
use them, getting an error back, and having to do another roundtrip.

This commit adds a new query parameter to the (`GET`) Ping endpoint
`/_ping?capabilities=1` that, if set, prompts the daemon to reply with a
JSON representation of it's capabilities

Since clients will generally already by hitting the `/_ping` endpoint in
order to negotiate API versions, this provides a way for clients to know
which features are supported without extra roundtrips.

This commit also adds a capability `registry-client-auth`, that
represents whether the engine supports the new client auth handling
protocol (which will only be supported, for now, if the containerd image
store integration is enabled) – without this, a client would have to
attempt to use the new features, get a "not implemented" error back, and
retry.

Signed-off-by: Laura Brehm <laurabrehm@hey.com>

corhere commented on api/swagger.yaml:
```diff
@@ -173,6 +173,17 @@ tags:
+  Capabilities:
+    description: "Represents engine capabilities"
+    type: "object"
+    required: ["registry-client-auth"]
```
I don't think it makes sense to declare any of the capability-flag properties as required. Any capability flags added after the first release would have to be optional as older daemons wouldn't be marshaling them. So it strikes me as odd to special-case the capabilities added in the initial implementation.

By declaring all capability properties as optional, we would be afforded more expressive power with the response, if we chose to make use of it. We could define that the presence of a capability with a false value signals that the daemon knows about the capability, despite it being disabled. As a corollary, the absence of a capability would therefore imply that the daemon is unaware of the capability at all.

laurazard commented on client/ping.go:
```diff
@@ -18,31 +24,39 @@ import (
 // for other non-success status codes, failing to connect to the API, or failing
 // to parse the API response.
-func (cli *Client) Ping(ctx context.Context) (types.Ping, error) {
+func (cli *Client) Ping(ctx context.Context, requestCapabilities bool) (types.Ping, error) {
```
Right, I meant to ask about that. I ended up adding a `client.PingWithOptions` and a `PingOptions` type – how does that sound?

laurazard commented on api/server/router/system/system_routes.go:
```diff
@@ -44,6 +44,22 @@ func (s *systemRouter) pingHandler(ctx context.Context, w http.ResponseWriter, r
 		return nil
 	}
+
+	if queryCapabilities := r.FormValue("capabilities"); queryCapabilities == "v1" {
```
I was looking into this, but [the link you shared](https://swagger.io/specification/v3/#discriminator-object) is for OpenAPI v3. In [OpenAPI (v2)](https://swagger.io/specification/v2/#discriminator-object), `discriminator` is just a `string` and the value **must be** the name of the schema of the payload.

While in OAS v3 we could do something like
```yaml
Capabilities:
    description: "Represents engine capabilities"
    type: "object"
    required:
    - "v_"
    properties:
      "v_":
        type: "integer"
        format: "uint16"
        description: ""
    discriminator:
      propertyName: "v_"
      mapping:
        1: CapabilitiesV1
  CapabilitiesV1:
    allOf:
    - $ref: '#/definitions/Capabilities'
    - type: object
      properties:
        registry-client-auth:
          description: "Whether the engine supports client auth handling"
          type: "boolean"
          x-nullable: true
    example:
      v_: 1
      registry-client-auth: true
```

In OAS v2 `discriminator` is only a `string` containing the name of the property used to discriminate, and the in the payload the value of that property **must match** the name of the schema of the payload, i.e. we'd have to do
```yaml
Capabilities:
    description: "Represents engine capabilities"
    type: "object"
    required:
    - "v_"
    properties:
      "v_":
        type: "integer"
        format: "uint16"
        description: ""
    discriminator: "v_"
  1:
    allOf:
    - $ref: '#/definitions/Capabilities'
    - type: object
      properties:
        registry-client-auth:
          description: "Whether the engine supports client auth handling"
          type: "boolean"
          x-nullable: true
    example:
      v_: 1
      registry-client-auth: true
```

which isn't as nice :'(

corhere replied:
There's a vendor extension you could use. https://redocly.com/docs/redoc/redoc-vendor-extensions#x-extendeddiscriminator

laurazard replied:
Oh, thank you! I've updated the Swagger spec to use that.

thaJeztah commented on api/swagger.yaml:
```diff
@@ -9709,6 +9719,83 @@ paths:
+          schema:
+            $ref: "#/definitions/Capabilities"
+          headers:
+            API-Version:
```
`Api-Version`; see https://github.com/moby/moby/pull/49054

thaJeztah commented on api/server/router/system/system_routes.go:
```diff
@@ -44,6 +47,28 @@ func (s *systemRouter) pingHandler(ctx context.Context, w http.ResponseWriter, r
+			if err != nil {
+				return err
+			}
+			_, err = w.Write(capabilitiesJson)
+			return err
```
This can probably use httputils.WriteJSON, which also handles setting the content-type etc;
https://github.com/moby/moby/blob/1e033e2d983df0f20caa8da5320dee3c3c2d7d63/api/server/router/container/container_routes.go#L70-L74

thaJeztah commented on api/server/router/system/system_routes.go:
```diff
@@ -44,6 +47,28 @@ func (s *systemRouter) pingHandler(ctx context.Context, w http.ResponseWriter, r
+		if err != nil {
+			return errdefs.InvalidParameter(err)
+		}
+		capabilities, err := s.handleCapabilities(ctx, wantedCapabilitiesVersion)
+		if err != nil {
+			log.G(ctx).WithError(err).Error("failed")
```
I need to jump out in a bit, but wondering if this should still fall through  to the plain-text "OK" response, or of this should return an error (will have to look closer)

laurazard replied:
Yeah, I was also wondering about that. I probably lean on that side, I feel like the client is specifically requesting capabilities, so if we fail the call should fail. OTOH, this seems less likely to cause problems, people put comments in daemon.json `features`, who knows if someone out there is putting random query params in their pings 😅

thaJeztah replied:
I went looking in what situations this would return an error, and it looks like currently there are none? (see other comment up/down (don't know where github will put it) 😂 

Somewhat considering that if things went wrong, it probably shouldn't ignore it. Or if we do, it could return an empty capabilities response (JSON, empty struct, or response with empty capabilities)

thaJeztah commented on daemon/capabilities/manager.go:
```diff
@@ -0,0 +1,120 @@
+// call to GetCapabilities will launch a goroutine to refresh the cache.
+// Subsequent calls to GetCapabilities will block until the cache has been
+// refreshed.
+func (m *Manager) GetCapabilities(ctx context.Context, p Provider, requestedVersion int) (VersionedCapabilities, error) {
```
Wondering; instead of passing the Provider as argument, would it make sense to pass it as option in the constructor, and keep a reference to it? i.e.;

```go
capabilities.NewManager(daemon)
```

After that, `GetCapabilities` can be called without having to pass a reference to the daemon. Doing so would probably also make sense from a perspective that the manager keeps a cache, which should be specific to the provider it used to look up things; passing a provider for each invocation would _technically_ mean I could swap it for any other provider, but the cache not taking that into consideration.

laurazard replied:
Yeah, this sounds good to me! It makes no sense to do it any other way – otherwise, a caller is just passing a provider to be used if the cache needs to be rebuilt, but if it doesn't it could be getting cache that was built from an entirely separate provider.

thaJeztah commented on api/server/router/system/system_routes.go:
```diff
@@ -57,6 +77,15 @@ func (s *systemRouter) swarmStatus() string {
 	return string(swarm.LocalNodeStateInactive)
 }
 
+func (s *systemRouter) handleCapabilities(ctx context.Context, capabilitiesVersion int) (capabilities.VersionedCapabilities, error) {
+	log.G(ctx).WithField("version", capabilitiesVersion).Debug("capabilities requested")
+	caps, err := s.backend.SystemCapabilities(ctx, capabilitiesVersion)
+	if err != nil {
+		return nil, err
+	}
+	return caps, nil
+}
+
```
Probably mostly a nit, but I was trying to see what this wrapper added as extra beyond the debug log;

- it looks like _currently_ `s.backend.SystemCapabilities` never returns an error
- ☝️ was this added to handle future errors for not-supported versions?
- if we don't have a good purpose yet; we could even consider dropping the error-return; the backend interfaces are mostly considered "internal" interfaces to decouple the API, so making signature changes there should be fine.

For the logs; possibly the existing logs capture some of that already (perhaps the "fetching system capabilities" could mention the version;

```console
DEBU[2024-12-13T15:19:59.621232216Z] capabilities requested                        spanID=6850a28621ab06d5 traceID=7367b3cf86d4c9e2d42e75f62eedc1ce version=123
DEBU[2024-12-13T15:19:59.621389882Z] fetching system capabilities                  spanID=6850a28621ab06d5 traceID=7367b3cf86d4c9e2d42e75f62eedc1ce
DEBU[2024-12-13T15:19:59.621546382Z] handling GET request                          method=GET module=api request-url="/_ping?capabilities=123" spanID=6850a28621ab06d5 status=200 traceID=7367b3cf86d4c9e2d42e75f62eedc1ce vars="map[]"
```

thaJeztah reviewed (commented):
Sorry for staggered comments dripping in; jumping between some things, so let me just post some blurbs I started jotting down :heart:

laurazard committed:
api/ping: add `types.PingOptions` to `client.Ping`

Signed-off-by: Laura Brehm <laurabrehm@hey.com>

laurazard commented on api/server/router/system/system_routes.go:
```diff
@@ -44,6 +47,28 @@ func (s *systemRouter) pingHandler(ctx context.Context, w http.ResponseWriter, r
+		if err != nil {
+			return errdefs.InvalidParameter(err)
+		}
+		capabilities, err := s.handleCapabilities(ctx, wantedCapabilitiesVersion)
```
Good catch, thanks!

thaJeztah commented on api/swagger.yaml:
```diff
@@ -9709,6 +9723,82 @@ paths:
+  # and https://github.com/OAI/OpenAPI-Specification/issues/182
+  /_ping​:
+    get:
+      summary: "/_ping?capabilities=v1"
```
Looks like we parse the argument now, so no longer allow a `v` prefix;

```suggestion
      summary: "/_ping?capabilities=1"
```

thaJeztah commented on api/swagger.yaml:
```diff
@@ -173,6 +173,20 @@ tags:
+    description: "Represents engine capabilities"
+    type: "object"
+    properties:
+      v_:
```
Looks like the field is `_v` (`_`  as prefix, not suffix)

thaJeztah commented on api/swagger.yaml:
```diff
@@ -173,6 +173,20 @@ tags:
+    properties:
+      v_:
+        description: "The version of capabilities this response represents."
+        type: integer
+        x-nullable: false
```
Also see other comment; it's good to put example values per-field, or at least it helps prevent diverging from the actual definition (so easier to maintain);


```suggestion
        type: integer
        x-nullable: false
        example: 1
```

thaJeztah commented on api/swagger.yaml:
```diff
@@ -173,6 +173,20 @@ tags:
     x-displayName: "System"
 
 definitions:
+  Capabilities:
+    description: "Represents engine capabilities"
+    type: "object"
+    properties:
+      v_:
+        description: "The version of capabilities this response represents."
+        type: integer
+        x-nullable: false
+      registry-client-auth:
+        description: "Whether the engine supports client auth handling."
+        type: "boolean"
+        x-nullable: true
+    example:
+      registry-client-auth: true
```
This one needs updating to match the new output. Where possible, it's good to put example-values _per field_ that way we make sure that the example output matches the actual structure. (Swagger will blindly show whatever is in `example`, even if that doesn't match the definition; putting example values per-field makes sure that every field we define is included in the example, and that it matches the schema);


⚠️ ❓ I noticed though that the `_v` field is embedded in the `capabilities` struct, and I wonder if it should be outside of that; otherwise `_v` could "technically" be considered a capability?


Currently it produces;

```bash
curl -s --unix-socket /var/run/docker.sock http://foo/_ping?capabilities=1
{"capabilities":{"_v":1,"registry-client-auth":false}}
```


And I wonder if that should be;


```bash
curl -s --unix-socket /var/run/docker.sock http://foo/_ping?capabilities=1
{"_v":1, "capabilities":{"registry-client-auth":false}}
```

laurazard replied:
I'd rather not do this – there's a lot of benefits to having it this way, i.e. OpenAPI support for discriminator fields, etc., as well as it makes sense for the type that holds those specific capabilities to include that version. Otherwise, the capabilities object itself does not contain enough information for a client to parse it. This only poses an issue if we want to treat capabilities as a map of arbitrary key-values, which we've already gone over and decided against.

laurazard replied:
Ended up changing this a bit (see [here](https://github.com/moby/moby/pull/49041#discussion_r1886922941) too) – I do think the current situation might be nicer (at least, UX-wise for this kind of stuff) but it does result in:
```go
curl -s --unix-socket /var/run/docker.sock http://foo/_ping?capabilities=1
{
	"capabilities": {
		"_v": 1,
		"capabilities": { 
			"registry-client-auth": false
		}
	}
}
```

laurazard replied:
From our last conversation, we ended up going with
```json
{
	"capabilities": {
		"_v": 1,
		"data": { 
			"registry-client-auth": false
		}
	}
}
```

thaJeztah commented on api/types/types.go:
```diff
@@ -8,6 +8,7 @@ import (
 	"github.com/docker/docker/api/types/image"
 	"github.com/docker/docker/api/types/swarm"
 	"github.com/docker/docker/api/types/volume"
+	"github.com/docker/docker/daemon/capabilities"
```
Wondering if we can reverse this; basically, we want to avoid anything inside `daemon/` to be imported outside of our repository, so perhaps we need (at least the "types" to be defined as a package in `api/xxx`.

laurazard replied:
Think so! Initially I had the capabilities logic in `daemon/`, so that wouldn't work because `daemon/` would have to import `api/` and that would cause issues because `api/` imports `daemon` somewhere down the chain, so I changed it. However, now that we have a capabilities package (which doesn't import the `daemon`) we can have it import `api`.

thaJeztah commented on daemon/capabilities/capabilities_versions.go:
```diff
@@ -0,0 +1,9 @@
+
+type capabilitiesV1 struct {
+	CapabilitiesBase
+	RegistryClientAuth bool `json:"registry-client-auth"`
```
Wondering if this is _too strict_. I was considering that the type we use (for the API response) could use a `map[string]XXX` (`bool` or  `any` ?), but we can define a list of capabilities (which can be versioned);


```go
type CapabilitiesBase struct {
	// CapabilitiesVersion is a natural number representing the version
	// of the capabilities API this response is for.
	//
	// This is part of the Capabilities version negotiation flow:
	// Since the API contract states that any client requesting
	// capabilities version N MUST also accept any capabilities
	// version M such that 1 < M < N, the daemon can pick a version
	// M = min(DAEMON_CAPABILITIES_VERSION, N), and send back a
	// response for that version.
	CapabilitiesVersion Version `json:"_v"`

	Capabilities map[string]any `json:"capabilities"`
}
```

This would preserve the option for 3-rd party capabilities (in the Docker Desktop, Mirantis MCR) to be added in their API proxy.

Not exactly sure what the best way is to document those capabilities in that case; perhaps something similar to how we document the runtime properties, where we document them as "well-known" ones, but in this case it may be slightly more strict, so perhaps not "well known" but a list of capabilities that may be returned 🤔 https://github.com/moby/moby/blob/01b98afeb65b4a5040d96958326fa04f5ab12c72/api/swagger.yaml#L6141-L6161

laurazard replied:
IMO, we can't have both:
- stable, solid version negotiation handling with backwards/forwards guarantees, and
- support for arbitrary values to be used by third parties

This also makes the response parsing logic on the client side significantly more complex. I'd rather 3rd party clients (which already must be aware of custom features added by their provider) parse the response themselves than explicitly support it.

If we really want to, I'd rather take an option on the client only to parse additional fields and return a `map[string]string`, or even to allow the client to pass in an `any` to be used for deserializing rather than change the types, which should be the canonical representation of what we actually support.

laurazard replied:
@corhere thoughts?

corhere replied:
I'm thinking we should reserve the capabilities map for upstream Moby and afford other ways to augment the ping response with arbitrary vendor extensions. That cleanly sidesteps any possibility of naming collisions between third-party capabilities and future Moby capabilities or third-party capabilities from another vendor. To that end, what if we versioned the ping response as a whole? In practical terms that would just be changing the query parameter name and flattening the JSON document by one level.

E.g.: `GET /ping?v=1` could return something along the lines of
```go
type PingBody struct {
        Version      uint                        `json:"_v"`
        Capabilities CapabilitiesV1              `json:"capabilities"`
        Extensions   map[string]json.RawMessage  `json:"extensions,omitempty"`
}
```

We could give vendors a sandbox to pass arbitrary data in the response which can easily be parsed out of the fixed structure and passed through the Engine API client library, with some guidelines (like naming their extension after a reversed DNS name they own) to avoid name collisions between vendors. We can therefore reserve all properties of the root JSON object for future non-breaking upstream additions to the ping response.

```json
{
  "_v": 1,
  "capabilities": {
    "registry-client-auth": true
  },
  "extensions": {
    "com.docker.desktop": {
      "capabilities-with-a-different-schema-because-why-not": ["foo", "bar", "!baz"],
      "hello": "world"
    }
  }
}
```

With the vendor extensions out of the way, the capabilities Go type can be defined as a struct rather than a map without issue as it does not need to be generic or extensible. The capability fields can be any type, and they can be strongly typed for each capability! It does not matter if the daemon sends a capability that the client doesn't know about; the client will silently ignore it.

laurazard replied:
This sounds good to me. @thaJeztah @vvoland does this sound good?

vvoland replied:
Looks good to me, one thing I'm still _meh_ about is the version being encoded in the JSON body itself.

On the client side, that forces the body to be unmarshalled twice (which isn't convenient, because we need to cache the response body).

Wondering, should we also (additionally) put that version in the HTTP header?

laurazard replied:
I think unmarshalling twice will always be the case – the `PingBody` type can't be tied to a specific version of capabilities (or any of it's other fields), because due to it's versioned nature those will always need to be unmarshalled according to the version:
```diff
type PingBody struct {
        Version      uint                        `json:"_v"`
-        Capabilities CapabilitiesV1             `json:"capabilities"`
+		 Capabilities json.RawMessage    	     `json:"capabilities"`
        Extensions   map[string]json.RawMessage  `json:"extensions,omitempty"`
}
```

I'm also thinking that, if the Ping response is versioned now, then the version negotiation should happen at the API level instead of passing the version down to https://github.com/laurazard/moby/blob/bb07daae222649fe33889e880325db291ab42db0/daemon/capabilities/manager.go#L39-L50 and managing it there, but in effect that just moves the whole `capabilities` package into the API layer, which makes it a bit harder to have the daemon call into it on a reload/config change due to import cycles iirc.

laurazard replied:
Oh, fair. We'd just unmarshal it into a `PingBodyV1`, or `PingBodyV2`, etc.?

vvoland replied:
Yeah, depending on the client's use case. Depending on how the struct changes in future, if all vN would be a superset of vN-1, you could just unmarshal to vN struct directly.

vvoland replied:
Anyway, I like Cory's proposal, my comment is just about additionally including the version in a HTTP header 😄

laurazard committed:
docs/api: document /_ping?capabilities=n endpoint

We've added a query parameter `capabilities` to the `/_ping` endpoint
that allows clients to request capabilities that the engine supports
in the `/_ping` response.

Document this change in `docs/api/version-history.md`, and update the
swagger yaml accordingly.

For Swagger, I've added a "separate" endpoint documenting the new
behavior. This has to do with Swagger not supporting different
responses for the same path (either based on query params or
content-types/other headers) – see [OAI/OpenAPI-Specification#146](https://github.com/OAI/OpenAPI-Specification/issues/146).

This is supported in OpenAPI 3.0
[(OAI/OpenAPI-Specification#146 (comment)](https://github.com/OAI/OpenAPI-Specification/issues/146#issuecomment-256648748)),
so we could fix this up in the future.

Signed-off-by: Laura Brehm <laurabrehm@hey.com>
